lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,629 @@
+---------------------------------------------------------------------------
+--
+--
+--
+--@DOC_wibox_widget_defaults_grid_EXAMPLE@
+-- @author getzze
+-- @copyright 2017 getzze
+-- @classmod wibox.layout.grid
+---------------------------------------------------------------------------
+
+local setmetatable = setmetatable
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+local table = table
+local pairs = pairs
+local ipairs = ipairs
+local math = math
+local util = require("awful.util")
+local base = require("wibox.widget.base")
+local beautiful = require("beautiful")
+
+
+local grid = { mt = {} }
+
+local properties = {
+                    "orientation", "superpose",
+                    "forced_num_rows", "forced_num_cols",
+                    "min_cols_size", "min_rows_size",
+                    --"max_cols_size", "max_rows_size",
+                   }
+
+local dir_properties = { "spacing", "padding", "fixed_size", "expand" }
+
+
+
+--- Reset a grid layout. This removes all widgets from the layout.
+-- **Signal:** widget::reset
+-- @param layout The layout you are modifying.
+-- @name reset
+-- @class function
+
+
+--- The preferred orientation for the grid.
+-- @beautiful beautiful.grid_orientation
+
+--- If widget superposition is allowed.
+-- @beautiful beautiful.grid_superpose
+
+--- The spacing between cells.
+-- @beautiful beautiful.grid_[horizontal|vertical]_spacing
+
+--- The space between the grid and the surrounding area.
+-- @beautiful beautiful.grid_[horizontal|vertical]_padding
+
+--- Controls if the columns/rows are expanded to use all the available width/height.
+-- @beautiful beautiful.grid_[horizontal|vertical]_expand
+
+--- Controls if the columns/rows have the same size or if the size depend on the content.
+-- @beautiful beautiful.grid_[horizontal|vertical]_fixed_size
```

Comment:
Since currently themes cannot just do this and setting this theme variable can have lots of unintended consequences, I would not recommend to add theme variables for `direction`. I'm also not sure that minimum size is a good idea...